crates/pathfinder/src/rpc/types.rs

```diff
@@ -655,58 +638,154 @@ pub mod reply {
+                            signature: Default::default(),
+                            nonce: TransactionNonce(Default::default()),
+                        },
+                        contract_address: txn.contract_address,
+                        contract_address_salt: txn.contract_address_salt,
+                        class_hash: txn.class_hash,
+                        constructor_calldata: txn.constructor_calldata.clone(),
+                    })
+                }
             }
         }
     }
 
     /// L2 transaction receipt as returned by the RPC API.
     #[serde_as]
-    #[skip_serializing_none]
     #[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
-    #[serde(deny_unknown_fields)]
-    pub struct TransactionReceipt {
-        pub txn_hash: StarknetTransactionHash,
-        #[serde_as(as = "Option<FeeAsHexStr>")]
-        #[serde(default)]
-        pub actual_fee: Option<Fee>,
-        pub status: TransactionStatus,
-        pub status_data: String,
+    #[serde(untagged)]
+    pub enum TransactionReceipt {
+        Invoke(InvokeTransactionReceipt),
+        // We can't differentiate between declare and deploy in an untagged enum: they
+        // have the same properties in the JSON.
+        DeclareOrDeploy(DeclareOrDeployTransactionReceipt),
+    }
+
+    impl TransactionReceipt {
+        pub fn hash(&self) -> StarknetTransactionHash {
+            match self {
+                Self::Invoke(tx) => tx.common.txn_hash,
+                Self::DeclareOrDeploy(tx) => tx.common.txn_hash,
+            }
+        }
+    }
+
+    #[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+    pub struct InvokeTransactionReceipt {
+        #[serde(flatten)]
+        pub common: CommonTransactionReceiptProperties,
+
         pub messages_sent: Vec<transaction_receipt::MessageToL1>,
-        #[serde(default)]
+        #[serde(default, skip_serializing_if = "Option::is_none")]
         pub l1_origin_message: Option<transaction_receipt::MessageToL2>,
         pub events: Vec<transaction_receipt::Event>,
     }
 
+    #[serde_as]
+    #[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+    pub struct CommonTransactionReceiptProperties {
+        pub txn_hash: StarknetTransactionHash,
+        #[serde_as(as = "FeeAsHexStr")]
+        pub actual_fee: Fee,
+        pub status: TransactionStatus,
+        #[serde(
+            default,
+            rename = "statusData",
+            skip_serializing_if = "Option::is_none"
+        )]
+        pub status_data: Option<String>,
+    }
+
+    #[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
+    pub struct DeclareOrDeployTransactionReceipt {
+        #[serde(flatten)]
+        pub common: CommonTransactionReceiptProperties,
+    }
+
     impl TransactionReceipt {
         pub fn with_status(
             receipt: sequencer::reply::transaction::Receipt,
             status: BlockStatus,
+            transaction: &sequencer::reply::transaction::Transaction,
```

Comment:
This coupling between the transaction and receipt makes me think we still haven't gotten the correct abstraction for transactions and receipts. I can't immedietely think of a better idea though.

Reply:
At some point I gave up thinking about this and proceeded just to have the stuff working.

Reply:
Yeah that's fair. One pain point at a time :)